File: impulse-kotlin/src/main/kotlin/org/impulsegraph/kotlin/ImpulseDsl.kt
package org.impulsegraph.kotlin

import org.impulsegraph.api.ArgType
import org.impulsegraph.api.ImpulseGraphQuery
import org.impulsegraph.api.ImpulseQueryBuilder
import org.impulsegraph.api.ReturnType

@DslMarker
annotation class ImpulseQueryDslMarker

/**
 * Type-safe, idiomatic Kotlin DSL builder for constructing Impulse Graph queries.
 */
@ImpulseQueryDslMarker
class ImpulseQueryDsl<R>(
    private val builder: ImpulseQueryBuilder<R> = ImpulseQueryBuilder()
) {

    /**
     * Define the input entity type and argument type for the query pipeline.
     */
    fun input(entityType: String, argType: ArgType = ArgType.SINGLE_NODE): ImpulseQueryDsl<R> {
        builder.input(entityType, argType)
        return this
    }

    /**
     * Add a CSR forward edge walk step over the specified relation name.
     */
    infix fun walkEdge(relationName: String): ImpulseQueryDsl<R> {
        builder.walkEdge(relationName)
        return this
    }

    /**
     * Walk edge with filter label.
     */
    fun walkEdgeFiltered(relationName: String, filterLabel: String): ImpulseQueryDsl<R> {
        builder.walkEdgeFiltered(relationName, filterLabel)
        return this
    }

    /**
     * Add a filtered CSR edge walk step based on numeric edge attribute comparison.
     */
    fun walkEdgeFilteredAttribute(relationName: String, attributeName: String, op: String, value: Double): ImpulseQueryDsl<R> {
        builder.walkEdgeFilteredAttribute(relationName, attributeName, op, value)
        return this
    }

    /**
     * Walk to target relation domain nodes.
     */
    infix fun walkTarget(relationName: String): ImpulseQueryDsl<R> {
        builder.walkTarget(relationName)
        return this
    }

    /**
     * Filter active candidate node set by comparing a node attribute against a numeric threshold.
     */
    fun filterNodeAttribute(attributeName: String, op: String, value: Double): ImpulseQueryDsl<R> {
        builder.filterNodeAttribute(attributeName, op, value)
        return this
    }

    /**
     * Add a map-reduce projection expression combining node and edge attributes.
     */
    fun projectExpression(nodeAttribute: String, operator: String, edgeAttribute: String): ImpulseQueryDsl<R> {
        builder.projectExpression(nodeAttribute, operator, edgeAttribute)
        return this
    }

    /**
     * Repeat a sub-query pipeline a fixed number of times.
     */
    fun repeat(count: Int, block: ImpulseQueryDsl<R>.() -> Unit): ImpulseQueryDsl<R> {
        builder.repeat({ subBuilder ->
            val dsl = ImpulseQueryDsl(subBuilder)
            dsl.block()
            subBuilder
        }, count)
        return this
    }

    /**
     * Repeat a sub-query pipeline until candidate set generation converges (reaches a fixed point).
     */
    fun repeatUntilStable(block: ImpulseQueryDsl<R>.() -> Unit): ImpulseQueryDsl<R> {
        builder.repeatUntilStable { subBuilder ->
            val dsl = ImpulseQueryDsl(subBuilder)
            dsl.block()
            subBuilder
        }
        return this
    }

    /**
     * Access domain-specific extended opcodes (ReBAC, PowerGrid island detection, motifs).
     */
    fun extended(block: ExtendedOpsDsl<R>.() -> Unit): ImpulseQueryDsl<R> {
        val extOps = builder.extended()
        val extDsl = ExtendedOpsDsl(extOps)
        extDsl.block()
        return this
    }

    /**
     * Terminal step: sum reduction over projected values.
     */
    fun <T> reduceSum(): ImpulseGraphQuery<T> = builder.reduceSum()

    /**
     * Terminal step: max reduction over projected values.
     */
    fun <T> reduceMax(): ImpulseGraphQuery<T> = builder.reduceMax()

    /**
     * Terminal step: min reduction over projected values.
     */
    fun <T> reduceMin(): ImpulseGraphQuery<T> = builder.reduceMin()

    /**
     * Terminal step: average reduction over projected values.
     */
    fun <T> reduceAvg(): ImpulseGraphQuery<T> = builder.reduceAvg()

    /**
     * Terminal step: first element reduction.
     */
    fun <T> reduceFirst(): ImpulseGraphQuery<T> = builder.reduceFirst()

    /**
     * Terminal collect step: materialize final result in the requested return type format.
     */
    fun <T> collect(returnType: ReturnType): ImpulseGraphQuery<T> = builder.collect(returnType)

    /**
     * Underlying Java query builder.
     */
    fun build(): ImpulseQueryBuilder<R> = builder
}

/**
 * Extended domain opcodes DSL wrapper.
 */
class ExtendedOpsDsl<R>(private val extOps: ImpulseQueryBuilder.ExtendedOps<R>) {
    fun islandDetect(src1Reg: Int, src2Reg: Int) {
        extOps.islandDetect(src1Reg, src2Reg)
    }

    fun rebacCheck(permission: String) {
        extOps.rebacCheck(permission)
    }

    fun motifMatch3() {
        extOps.motifMatch3()
    }
}

/**
 * Top-level idiomatic builder function for constructing an Impulse Graph query pipeline.
 */
inline fun <R> impulseQuery(block: ImpulseQueryDsl<R>.() -> Unit): ImpulseQueryDsl<R> {
    val dsl = ImpulseQueryDsl<R>()
    dsl.block()
    return dsl
}

File: impulse-scala/src/main/scala/org/impulsegraph/scala/ImpulseDsl.scala
package org.impulsegraph.scala

import org.impulsegraph.api.{ArgType, ImpulseGraphQuery, ImpulseQueryBuilder, ReturnType}
import scala.annotation.targetName

class ImpulseQueryDsl[R](private val builder: ImpulseQueryBuilder[R] = new ImpulseQueryBuilder[R]()):

  def input(entityType: String, argType: ArgType = ArgType.SINGLE_NODE): ImpulseQueryDsl[R] =
    builder.input(entityType, argType)
    this

  def input(entityType: String, seed: NodeId): ImpulseQueryDsl[R] =
    builder.input(entityType, ArgType.SINGLE_NODE)
    this

  infix def walkEdge(relationName: String): ImpulseQueryDsl[R] =
    builder.walkEdge(relationName)
    this

  @targetName("walkEdgeRel")
  infix def walkEdge(relationName: RelationName): ImpulseQueryDsl[R] =
    builder.walkEdge(relationName.value)
    this

  def walkEdgeFiltered(relationName: String, filterLabel: String): ImpulseQueryDsl[R] =
    builder.walkEdgeFiltered(relationName, filterLabel)
    this

  @targetName("walkEdgeFilteredRel")
  def walkEdgeFiltered(relationName: RelationName, filterLabel: String): ImpulseQueryDsl[R] =
    builder.walkEdgeFiltered(relationName.value, filterLabel)
    this

  def walkEdgeFilteredAttribute(
      relationName: String,
      attributeName: String,
      op: String,
      value: Double
  ): ImpulseQueryDsl[R] =
    builder.walkEdgeFilteredAttribute(relationName, attributeName, op, value)
    this

  @targetName("walkEdgeFilteredAttributeRel")
  def walkEdgeFilteredAttribute(
      relationName: RelationName,
      attributeName: String,
      op: String,
      value: Double
  ): ImpulseQueryDsl[R] =
    builder.walkEdgeFilteredAttribute(relationName.value, attributeName, op, value)
    this

  infix def walkTarget(relationName: String): ImpulseQueryDsl[R] =
    builder.walkTarget(relationName)
    this

  @targetName("walkTargetRel")
  infix def walkTarget(relationName: RelationName): ImpulseQueryDsl[R] =
    builder.walkTarget(relationName.value)
    this

  def filterNodeAttribute(attributeName: String, op: String, value: Double): ImpulseQueryDsl[R] =
    builder.filterNodeAttribute(attributeName, op, value)
    this

  def projectExpression(nodeAttribute: String, operator: String, edgeAttribute: String): ImpulseQueryDsl[R] =
    builder.projectExpression(nodeAttribute, operator, edgeAttribute)
    this

  def repeat(count: Int)(fn: ImpulseQueryDsl[R] ?=> Unit): ImpulseQueryDsl[R] =
    builder.repeat(
      subBuilder => {
        val subDsl = new ImpulseQueryDsl[R](subBuilder)
        fn(using subDsl)
        subBuilder
      },
      count
    )
    this

  def repeatUntilStable(fn: ImpulseQueryDsl[R] ?=> Unit): ImpulseQueryDsl[R] =
    builder.repeatUntilStable(subBuilder => {
      val subDsl = new ImpulseQueryDsl[R](subBuilder)
      fn(using subDsl)
      subBuilder
    })
    this

  def extended(fn: ExtendedOpsDsl[R] ?=> Unit): ImpulseQueryDsl[R] =
    val extOps = builder.extended()
    val extDsl = new ExtendedOpsDsl[R](extOps)
    fn(using extDsl)
    this

  def reduceSum[T](): ImpulseGraphQuery[T] = builder.reduceSum[T]()
  def reduceMax[T](): ImpulseGraphQuery[T] = builder.reduceMax[T]()
  def reduceMin[T](): ImpulseGraphQuery[T] = builder.reduceMin[T]()
  def reduceAvg[T](): ImpulseGraphQuery[T] = builder.reduceAvg[T]()
  def reduceFirst[T](): ImpulseGraphQuery[T] = builder.reduceFirst[T]()

  def collect[T](returnType: ReturnType): ImpulseGraphQuery[T] = builder.collect[T](returnType)

  def build(): ImpulseQueryBuilder[R] = builder

class ExtendedOpsDsl[R](private val extOps: ImpulseQueryBuilder.ExtendedOps[R]):
  def islandDetect(src1Reg: Int, src2Reg: Int): Unit = extOps.islandDetect(src1Reg, src2Reg)
  def rebacCheck(permission: String): Unit = extOps.rebacCheck(permission)
  def motifMatch3(): Unit = extOps.motifMatch3()

// Context Functions for receiver-less Scala 3 DSL syntax
def input(entityType: String, argType: ArgType = ArgType.SINGLE_NODE)(using dsl: ImpulseQueryDsl[?]): Unit =
  dsl.input(entityType, argType)

def input(entityType: String, seed: NodeId)(using dsl: ImpulseQueryDsl[?]): Unit =
  dsl.input(entityType, seed)

def walkEdge(relationName: String)(using dsl: ImpulseQueryDsl[?]): Unit =
  dsl.walkEdge(relationName)

@targetName("walkEdgeRel")
def walkEdge(relationName: RelationName)(using dsl: ImpulseQueryDsl[?]): Unit =
  dsl.walkEdge(relationName)

def walkTarget(relationName: String)(using dsl: ImpulseQueryDsl[?]): Unit =
  dsl.walkTarget(relationName)

@targetName("walkTargetRel")
def walkTarget(relationName: RelationName)(using dsl: ImpulseQueryDsl[?]): Unit =
  dsl.walkTarget(relationName)

def walkEdgeFiltered(relationName: String, filterLabel: String)(using dsl: ImpulseQueryDsl[?]): Unit =
  dsl.walkEdgeFiltered(relationName, filterLabel)

@targetName("walkEdgeFilteredRel")
def walkEdgeFiltered(relationName: RelationName, filterLabel: String)(using dsl: ImpulseQueryDsl[?]): Unit =
  dsl.walkEdgeFiltered(relationName, filterLabel)

def walkEdgeFilteredAttribute(
    relationName: String,
    attributeName: String,
    op: String,
    value: Double
)(using dsl: ImpulseQueryDsl[?]): Unit =
  dsl.walkEdgeFilteredAttribute(relationName, attributeName, op, value)

@targetName("walkEdgeFilteredAttributeRel")
def walkEdgeFilteredAttribute(
    relationName: RelationName,
    attributeName: String,
    op: String,
    value: Double
)(using dsl: ImpulseQueryDsl[?]): Unit =
  dsl.walkEdgeFilteredAttribute(relationName, attributeName, op, value)

def filterNodeAttribute(attributeName: String, op: String, value: Double)(using dsl: ImpulseQueryDsl[?]): Unit =
  dsl.filterNodeAttribute(attributeName, op, value)

def projectExpression(nodeAttribute: String, operator: String, edgeAttribute: String)(using
    dsl: ImpulseQueryDsl[?]
): Unit =
  dsl.projectExpression(nodeAttribute, operator, edgeAttribute)

def repeat(count: Int)(fn: ImpulseQueryDsl[?] ?=> Unit)(using dsl: ImpulseQueryDsl[?]): Unit =
  dsl.asInstanceOf[ImpulseQueryDsl[Any]].repeat(count)(fn)

def repeatUntilStable(fn: ImpulseQueryDsl[?] ?=> Unit)(using dsl: ImpulseQueryDsl[?]): Unit =
  dsl.asInstanceOf[ImpulseQueryDsl[Any]].repeatUntilStable(fn)

def extended(fn: ExtendedOpsDsl[?] ?=> Unit)(using dsl: ImpulseQueryDsl[?]): Unit =
  dsl.asInstanceOf[ImpulseQueryDsl[Any]].extended(fn)

def rebacCheck(permission: String)(using ext: ExtendedOpsDsl[?]): Unit =
  ext.rebacCheck(permission)

def islandDetect(src1Reg: Int, src2Reg: Int)(using ext: ExtendedOpsDsl[?]): Unit =
  ext.islandDetect(src1Reg, src2Reg)

def motifMatch3()(using ext: ExtendedOpsDsl[?]): Unit =
  ext.motifMatch3()

def impulseQuery[R](fn: ImpulseQueryDsl[R] ?=> Unit): ImpulseQueryDsl[R] =
  val dsl = new ImpulseQueryDsl[R]()
  fn(using dsl)
  dsl
